### ISSUE TITLE
a problem about ground-truth center-ness generation for inference

### ISSUE THREAD

**Issue by gaoyuchris** [tianzhi0549/FCOS#32]
I have tried using the ground-truth center-ness for inference, however only got  AP 38.8 rather than 42.1 metioned in the paper.  I wonder if there are some errors in my code of ground-truth center-ness generation.  Can you help me find the problem or give me your code? thx. 
My code is as follow:
````
        labels, reg_targets = self.prepare_targets(locations, targets)
        centerness_target = []
        for l in range(len(labels)):
            reg_targets_flatten = reg_targets[l].reshape(-1, 4)
            # import pdb;pdb.set_trace()
            reg_targets_flatten = (labels[l]!=0)[:,None].float()*reg_targets_flatten
            reg_targets_flatten = self.compute_centerness_targets(reg_targets_flatten)
            centerness_target.append(reg_targets_flatten.reshape(centerness[l].shape))

        return centerness_target
```

**Comment by tianzhi0549** [comment]
@gaoyuchris Can you try to change the order of the two lines?
```
reg_targets_flatten = (labels[l]!=0)[:,None].float()*reg_targets_flatten
reg_targets_flatten = self.compute_centerness_targets(reg_targets_flatten)
```

**Comment by gaoyuchris** [comment]
@tianzhi0549 Thank you for your reply. I have tried,  but got the same AP with 38.8

**Comment by tianzhi0549** [comment]
@gaoyuchris is it convenient for you to post all the changes you made here?

**Comment by gaoyuchris** [comment]
@tianzhi0549 
I just add the code in line 85 "FCOS/maskrcnn_benchmark/modeling/rpn/fcos/loss.py" for some images have zero target 
```
if len(bboxes)==0:
        labels.append(torch.zeros(len(xs), dtype=torch.int64, device=xs.device))
        reg_targets.append(torch.zeros([len(xs), 4], device=xs.device))
        continue
```
and use the code 
```
labels, reg_targets = self.prepare_targets(locations, targets)
centerness_target = []
for l in range(len(labels)):
        reg_targets_flatten = reg_targets[l].reshape(-1, 4)
        reg_targets_flatten = self.compute_centerness_targets(reg_targets_flatten)
        reg_targets_flatten = (labels[l]!=0).float()*reg_targets_flatten
        centerness_target.append(reg_targets_flatten.reshape(centerness[l].shape))
return centerness_target
```

for the function I change as follow for the reason that some points not in the bboxes will get centerness<0:
```
def compute_centerness_targets(self, reg_targets):
        left_right = reg_targets[:, [0, 2]]
        top_bottom = reg_targets[:, [1, 3]]
        centerness = (left_right.min(dim=-1)[0] / (left_right.max(dim=-1)[0]+1e-10)) * \
                      (top_bottom.min(dim=-1)[0] / (top_bottom.max(dim=-1)[0]+1e-10))
        centerness[centerness<0]=0.0
        res = torch.sqrt(centerness)
        return res
```

**Comment by tianzhi0549** [comment]
@gaoyuchris How do you use the computed ground-truth center-ness?

**Comment by gaoyuchris** [comment]
@tianzhi0549 I used the computed ground-truth center-ness in focs.py as follows:
```
def _forward_test(self, locations, box_cls, box_regression, centerness, image_sizes, targets):
        centerness_target = self.loss_evaluator.get_centerness_target(
            locations, box_cls, box_regression, centerness, targets
        )
        boxes = self.box_selector_test(
            locations, box_cls, box_regression, 
            centerness_target, image_sizes
        )

        return boxes, {}
```
the" self.loss_evaluator.get_centerness_target(locations, box_cls, box_regression, centerness, targets)" is the func mentioned before.
